### PR TITLE
feat/automapper post

### PR DESCRIPTION
**Pull request by ilaborde**

### CODE REVIEW

File: posts/ilaborde/1.automapper.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+---
+title: EF Core Simplify your queries with AutoMapper
+published: false
```
Comment (navarroaxel) [inline posts/ilaborde/1.automapper.md:3]: Set to `true` to publish on merge
```suggestion
published: true
```

File: posts/ilaborde/1.automapper.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    public class PostSummaryDto
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public int CommentsCount { get; set; }
+    }
+```
+
+When querying data with Entity Framework is very common to query data using the Include method, and then manually map the collection of posts to a collection of PostDto:
```
Comment (navarroaxel) [inline posts/ilaborde/1.automapper.md:61]:
```suggestion
When querying data with Entity Framework is very common to query data using the Include method, and then manually map the collection of posts to a collection of `PostDto`:
```

File: posts/ilaborde/1.automapper.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+```SQL
+    SELECT [p].[Id], [p].[Title], [b].[Id], [b].[Name], [c].[Id]
+    FROM [Posts] AS [p]
+    INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
+    LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+    ORDER BY [p].[Id], [b].[Id]
+```
+
+_The .ProjectTo<PostDto>() will tell AutoMapper’s mapping engine to emit a select clause to the IQueryable that will inform Entity Framework that it only needs to query the fields presents in the DTO._
```
Comment (navarroaxel) [inline posts/ilaborde/1.automapper.md:141]:
```suggestion
_The `.ProjectTo<PostDto>()` will tell AutoMapper’s mapping engine to emit a select clause to the IQueryable that will inform Entity Framework that it only needs to query the fields presents in the DTO._
```

File: posts/ilaborde/1.automapper.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+```c#
+    public class PostSummaryDto
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public int CommentsCount { get; set; }
+    }
+```
+
+And writing again only 1 C# line:
```
Comment (navarroaxel) [inline posts/ilaborde/1.automapper.md:154]: Sounds better?
```suggestion
And writing again only 1 line on C#:
```

File: posts/ilaborde/1.automapper.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
+---
```
Comment (navarroaxel) [inline posts/ilaborde/1.automapper.md:1]: Rename file to `01.automapper.md` to have a good order when you have more than 9 posts in the future.

File: posts/ilaborde/01.automapper.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    }
+```
+
+When querying data with Entity Framework is very common to query data using the Include method, and then manually map the collection of posts to a collection of `PostDto`:
+
+```c#
+  var posts = dbContext.Posts
+      .Include(x => x.Blog)
+      .Include(x => x.Comments)
+      .ToListAsync();
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:67]:
```suggestion
var posts = dbContext.Posts
    .Include(x => x.Blog)
    .Include(x => x.Comments)
    .ToListAsync();
```

File: posts/ilaborde/01.automapper.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+    ORDER BY [p].[Id], [b].[Id]
+```
+
+### ProjectTO to the rescue
+
+AutoMapper has a simple way to solve this:
+
+```c#
+   var posts = mapper.ProjectTo<PostDto>(dbContext.Posts, null).ToListAsync();
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:128]:
```suggestion
 var posts = mapper.ProjectTo<PostDto>(dbContext.Posts, null).ToListAsync();
```

File: posts/ilaborde/01.automapper.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+The common case for the `ProjectTo` is using small DTOs/models. e.g:
+
+```c#
+    public class PostSummaryDto
+    {
+        public int Id { get; set; }
+        public string Title { get; set; }
+        public int CommentsCount { get; set; }
+    }
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:151]:
```suggestion
public class PostSummaryDto
{
    public int Id { get; set; }
    public string Title { get; set; }
    public int CommentsCount { get; set; }
}
```

File: posts/ilaborde/01.automapper.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+```
+
+And writing again only 1 line on C#:
+
+```c#
+    var postSummary = mapper.ProjectTo<PostSummaryDto>(dbContext.Posts, null).ToListAsync();
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:157]:
```suggestion
var postSummary = mapper.ProjectTo<PostSummaryDto>(dbContext.Posts, null).ToListAsync();
```

File: posts/ilaborde/01.automapper.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
+---
+title: EF Core Simplify your queries with AutoMapper
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:2]:
```suggestion
title: Entity Framework: Simplify your queries with AutoMapper
```

File: posts/ilaborde/01.automapper.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+published: true
+description: Simplify EF core queries with AutoMapper
+tags: 'net, entity-framework, automapper'
+cover_image: ./assets/1-automapper-cover.png
+---
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:7]:
```suggestion
id: 701321
---
```

File: posts/ilaborde/01.automapper.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
```
Comment (pabloromeo) [inline posts/ilaborde/01.automapper.md:8]: An introduction is missing, in order to give context to the reader

File: posts/ilaborde/01.automapper.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+```SQL
+    SELECT [p].[Id], [p].[Title], [b].[Id], [b].[Name], [c].[Id]
+    FROM [Posts] AS [p]
+    INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
+    LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+    ORDER BY [p].[Id], [b].[Id]
+```
+
+### ProjectTO to the rescue
```
Comment (pabloromeo) [inline posts/ilaborde/01.automapper.md:123]: It says "to the rescue" of some problem, but the problem was never actually mentioned. For example, in the first query you show, you probably want to mention that full entities are meterialized, including properties you might not even need.
The second alternative with a manual projection would solve that, but what problems does it have using it that way? THAT is what projecTo solves and justifies the "to the rescue".

File: posts/ilaborde/01.automapper.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+      .ToListAsync();
+  var collectionOfPostDto = mapper.Map<PostDto>(posts);
+```
+
+Obviously generating the same SQL query.
+
+You can do the same using an Entity Framework projection, just picking the fields that you need:
+
+```c#
+    var post = dbContext.Posts
```
Comment (pabloromeo) [inline posts/ilaborde/01.automapper.md:98]:
```suggestion
    var post = await dbContext.Posts
```

File: posts/ilaborde/01.automapper.md (hunk 7 above)
Comment (pabloromeo) [inline posts/ilaborde/01.automapper.md:128]:
```suggestion
   var posts = await mapper.ProjectTo<PostDto>(dbContext.Posts, null).ToListAsync();
```

File: posts/ilaborde/01.automapper.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    SELECT [p].[Id], [p].[Title], [b].[Id], [b].[Name], [c].[Id]
+    FROM [Posts] AS [p]
+    INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
+    LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+    ORDER BY [p].[Id], [b].[Id]
+```
+
+### ProjectTO to the rescue
+
+AutoMapper has a simple way to solve this:
```
Comment (pabloromeo) [inline posts/ilaborde/01.automapper.md:125]: The more common usage is applying it as an extension method to an IQueriable.

File: posts/ilaborde/01.automapper.md (hunk 7 above)
Comment (pabloromeo) [inline posts/ilaborde/01.automapper.md:128]:
```suggestion
   var posts = await mapper.ProjectTo<PostDto>(dbContext.Posts, null).ToListAsync();
```

**Review by pabloromeo — CHANGES_REQUESTED** [review]

File: posts/ilaborde/01.automapper.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
+---
+title: 'Entity Framework Core: Simplify your queries with AutoMapper'
+published: true
+description: We can create smaller and performant queries using Entity Framework Core and AutoMapper.
+tags: 'net, entity-framework, automapper'
+cover_image: ./assets/01-automapper-cover.png
+---
+
+For the examples of code, we are going to use the following entities.
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:9]:
```suggestion
For the examples of code, we are going to use the following entities:
```

File: posts/ilaborde/01.automapper.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+public class Comment
+{
+    public int Id { get; set; }
+    public string Content { get; set; }
+    public Post Post { get; set; }
+}
+```
+
+And DTOs
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:35]: Same punctuation like above
```suggestion
And DTOs:
```

File: posts/ilaborde/01.automapper.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+When querying data with Entity Framework is very common to use the Include method. But that can have two drawbacks: you will query fields that probably you don't need, and the second one maybe you will need to map your entities to another model or DTO. 
+
+```c#
+var posts = await dbContext.Posts
+    .Include(x => x.Blog)
+    .Include(x => x.Comments)
+    .ToListAsync();
+
+// map to DTO
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:69]:
```suggestion
// map to DTO
```

File: posts/ilaborde/01.automapper.md [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+    .ToListAsync();
+
+// map to DTO     
+```
+
+For this query Entity Framework will generate the following SQL code:
+
+```SQL
+SELECT [p].[Id], [p].[BlogId], [p].[Title], [b].[Id], [b].[Name], 
+[c].[Id], [c].[Content], [c].[PostId]
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:76]:
```suggestion
    [c].[Id], [c].[Content], [c].[PostId]
```

File: posts/ilaborde/01.automapper.md [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+SELECT [p].[Id], [p].[Title], [b].[Id], [b].[Name], [c].[Id]
+FROM [Posts] AS [p]
+INNER JOIN [Blogs] AS [b] ON [p].[BlogId] = [b].[Id]
+LEFT JOIN [Comments] AS [c] ON [p].[Id] = [c].[PostId]
+ORDER BY [p].[Id], [b].[Id]
+```
+
+_The `.ProjectTo<PostDto>()` will tell AutoMapper's mapping engine to emit a select clause to the IQueryable that will inform Entity Framework that it only needs to query the fields presents in the DTO._
+
+A common case for the `ProjectTo` is using small DTOs/models. For example, we can have a smaller DTO to just return a summary of the posts to list in the frontend, with just the id, name, and comments counts.
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:133]:
```suggestion
A common case for the `.ProjectTo` is using small DTOs/models. For example, we can have a smaller DTO to just return a summary of the posts to list in the frontend, with just the id, name, and comments counts. 
```

File: posts/ilaborde/01.automapper.md [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+var postSummary = await mapper.ProjectTo<PostSummaryDto>(dbContext.Posts, null).ToListAsync();
+```
+
+The SQL code looks really nice:
+
+```SQL
+SELECT [p].[Id], [p].[Title], (
+    SELECT COUNT(*)
+    FROM [Comments] AS [c]
+    WHERE [p].[Id] = [c].[PostId]) AS [CommentsCount]
```
Comment (navarroaxel) [inline posts/ilaborde/01.automapper.md:156]: It's hard to read when the sub-select ends
```suggestion
        SELECT COUNT(*)
        FROM [Comments] AS [c]
        WHERE [p].[Id] = [c].[PostId]
    ) AS [CommentsCount]
```

**Review by pabloromeo — APPROVED** [review]

**Review by navarroaxel — APPROVED** [review]

**Pull request merged**